rfcs/2021-11-12-9811-vrl-vm.md

```diff
@@ -0,0 +1,347 @@
+
+## Scope
+
+### In scope
+
+This RFC is purely about developing a enum VM for VRL. It will discuss the
+risks involved in running VRL with a VM and the measures we will take to
+mitigate those risks.
+
+### Out of scope
+
+Any other performance issues relating to VRL will be discussed in
+[rfcs/2021-10-14-9811-vrl-performance.md](https://github.com/vectordotdev/vector/pull/9812)
+
+## Pain
+
+Vrl compiles to an AST that is then walked during resolution. Each node in that
+tree is boxed and stored in disparate regions of memory. As a result walking
+the tree means that the CPU caches must be constantly swapped.
+
+## Proposal
+
+### Implementation
+
+Instead we can create an enum VM to store the execution of the Vrl program.
+
+The enum is essentially a big enum of instructions:
+
+```rust
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub enum OpCode {
+    Return,
+    Constant,
+    Negate,
+    Add,
+    Subtract,
+    Multiply,
+    Divide,
+    Print,
+    Not,
+    Greater,
+    GreaterEqual,
+    Less,
+    LessEqual,
+    NotEqual,
+    Equal,
+    Pop,
+    JumpIfFalse,
+    Jump,
+    SetPath,
+    GetPath,
+    Call,
+    ...
+}
+
+enum Instruction {
+    Opcode(Opcode),
+    Literal(LiteralIndex),
+}
+
+pub struct LiteralIndex(usize);
+```
+
+The Vm is a struct comprising of the following fields:
+
+```rust
+#[derive(Clone, Debug, Default)]
+pub struct Vm {
+    instructions: Vec<Instruction>,
+    constants: Vec<Literal>,
+    targets: Vec<Variable>,
+    stack: Vec<Value>,
+    parameter_stack: Vec<Option<Value>>,
+    error: Option<Error>,
+    instruction_pointer: usize,
+}
+```
+
+- instructions
+
+The instructions field is a `Vec` of `Instruction`. An instruction can be
+either an OpCode or some data for that OpCode. For example, the instructions
+`[.., OpCode(Constant), Literal(12), ..]` when evaluated will load the constant
+stored in the `values` `Vec` that is found in position 12 onto the stack.
+
+- constants
+
+A list of constant values found in the program. Since the bytecode only
+contains integers any actual values must be stored here. This also allows
+literals to be deduped.
+
+- targets
+
+A list of paths used in the program, similar to `values`.
+
+- stack
+
+The Vm is a stack based Vm. Every expression that is evaluated pushes the
+result on the stack. Every operation pulls the values it uses from the stack.
+
+- parameter_stack
+
+For function calls, the parameters need to be evaluated and passed to the
+function. The stack is a `Vec` of parameter name and value. We need an OpCode
+that will copy a value from the main stack to the parameter stack tagged with
+the parameter name. The `FunctionArgs` compiler will dump this OpCode after the
+code to evaluate the given arg is dumped.
+
+- error
+
+If an operation errors, it needs to populate this field. Operations such as
+the error assignment (`result, error = thing()`) will need to check if this
+field has been populated and action accordingly.
```

Comment:
If we have a value stack, should this value not also go on the stack, or is the error value really special? What opcode handles its value?

Reply:
The error value is a special value in that it is an Error and needs to be identified as such. 

We could make the stack a `Vec<Result<Value, Error>>`, but I think that would be more of a pain than just storing the errors elsewhere - similar to how `C` uses `errno`.

Reply:
Yeah, that's going to be a big pain. I was thinking more that functions that emit the pair of results would push both on the stack, and callers that don't use one or the other simply pop the relevant entry into oblivion. This is likely simpler to implement for the spike, though.